### PR TITLE
[inductor] Parallelize Max Autotune step 2: Use all GPUs

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107983
* #107982

Summary: Step 2 in revamping subprocess autotune to support multiple GPUs: use a pool of subprocesses and distribute benchmark calls across them.

Test Plan:
`python test/inductor/test_max_autotune.py`
`TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1 TORCHINDUCTOR_MAX_AUTOTUNE=1 python benchmarks/dynamo/torchbench.py --device cuda --performance --backend inductor --inference --only hf_Bart`
`TORCHINDUCTOR_AUTOTUNE_MULTI_DEVICE=1 TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1 TORCHINDUCTOR_MAX_AUTOTUNE=1 python benchmarks/dynamo/torchbench.py --device cuda --performance --backend inductor --inference --only hf_Bart`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Looks like you have test failures. We might need to turn this off by default for now, or only on during aot_compilation. can do that in subsequent pr

File: torch/_inductor/autotune_process.py
Comment (masnesral) [inline torch/_inductor/autotune_process.py:1]: This is to support the type hint `Queue[TuningProcess]` in the 3.8 builds since apparently a subscriptable Queue type was added later.

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [inline torch/_inductor/select_algorithm.py:764]: This part probably needs some explanation (and possibly some input from anyone who's familiar with this code). In the existing implementation, this error handling would seem to apply to both the in-process and sub-process benchmarking paths. However, the sub-process path was never raising any exceptions in the first place -- it's only providing the latency result or "inf" if the sub-process exits unexpectedly. So I moved this error handling to the in-process path to be more honest about what can happen. 

It's not clear to me whether these conditions can happen in the sub-process path or whether they always applied to the in-process path only. (I never saw them in my testing). If these conditions are possible for both in-process and sub-process, and we want to capture them for both paths, then I'd probably want to refactor along these lines:
a) The sub-process logic should capture exceptions and communicate them back to the parent
b) All relevant functions operate on lists of choices and corresponding lists of results

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [inline torch/_inductor/select_algorithm.py:852]: diff looks a little weird here too. The benchmark_in_sub_process impl is below.

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [inline torch/_inductor/select_algorithm.py:860]: The benchmarking all happens in parallel now, so I didn't think the timestamp info was as relevant.

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:145]: add class docstring ?

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:158]: we should set count so it is no greater than `torch.cuda.device_count()`

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:182]: this can be a top level import right

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:222]: is there any worry here about putting back in the same process after it had failed ? say, if the process has been corrupted ? might be safer to just exit and start new process. what does old code do ?

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:366]: sorry, why do we need to do this ? would be nice to keep separation between test and source code if possible.

One idiom that might be helpful is runtime patching objects in test code, e.g. `torch._inductor.autotune_process.BenchmarkRunner = TestBenchmarkRequest`

File: torch/_inductor/select_algorithm.py
Comment (eellison) [reply torch/_inductor/select_algorithm.py:764]: cc @shunting314

File: torch/_inductor/select_algorithm.py
Comment (eellison) [inline torch/_inductor/select_algorithm.py:877]: is there a reason for this ?

**Review by eellison — COMMENTED** [review]
cool ! tagging @shunting314 who did a lot of the initial impl

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:182]: @eellison do you just mean to move it out of the class or do I misunderstand? I stuck with the pattern already established. I assume it was done this way to avoid the atexit registration if we never actually initialize and use the sub-process path?

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:222]: @eellison this is related to a comment you made on the parent diff. The way I tried to organize things, it's the TuningProcess class that manages its subprocess and always ensures that it's running. So it should be safe as long as the TuningProcess is written correctly to do that.

The old code always calls "initialize" from the outside the TuningProcess before using it, and it's that initialize method that checks if the process previously crashed and restarts it if so. I think the abstraction is cleaner if we just contain that logic inside the TuningProcess. One suggestion I made on the other diff was that it might be clearer to the reader if we optionally restart sub-processes before sending a message (instead, I did it immediately after crash). Either approach is fine, I think.

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:366]: @eellison, sure lemme try your idiom and see if that solves my problem. My guess is that it won't, tho. The problem that I was running into was that the sub-process didn't know how to unpickle a `TestBenchmarkRequest` if that class is defined in the test file (which is unknown to this file). I think we'd still have that problem. Another option is to define the Test class in the test file (where it belongs), but simply import it here, along with a lint directive to ignore and not remove that import, perhaps? That struck me as equally hacky. Yet another option is dispense with this special test class altogether and manipulate real BenchmarkRequest objects in the unit test. That's how I started, but the logic was so much more convoluted than the test implementation you see in this diff -- I liked the simplicity of the test. But I can definitely try harder to get a tidy-looking test without putting this here if you prefer that route.

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [reply torch/_inductor/select_algorithm.py:877]: @eellison I don't know. Maybe @shunting314 can help answer this part too? I maintained this part of the logic from the existing implementation. If there's no real technical reason why we can't use the sub-process path for both, then I'm game to upgrade to full sub-process. Though perhaps as a next step? There will be at least some minor refactoring necessary because the in-process vs. sub-process paths look like they operate on different types.

**Comment by shunting314** [comment]
Can you also share how much speedup we can get by leveraging multi-GPU. I actually have a similar PR sometime back https://github.com/pytorch/pytorch/pull/96807/files to extend the sub-process autotuning to multi-process/multi-GPU autotuning . We didn't push forward at that time because people thought a global cache is more promising.

File: torch/_inductor/autotune_process.py
Comment (eellison) [reply torch/_inductor/autotune_process.py:182]: Yea, nevermind.

File: torch/_inductor/select_algorithm.py
Comment (eellison) [reply torch/_inductor/select_algorithm.py:877]: Agreed, it's not necessary to do immediately.

File: torch/_inductor/select_algorithm.py
Comment (shunting314) [reply torch/_inductor/select_algorithm.py:860]: Hmm, it would still be helpful to know how long each benchmark request takes to be processed?

File: torch/_inductor/select_algorithm.py
Comment (shunting314) [inline torch/_inductor/select_algorithm.py:851]: nit: should we uniformly call it 'choice' and 'choices'?

File: torch/_inductor/select_algorithm.py
Comment (shunting314) [reply torch/_inductor/select_algorithm.py:764]: yea, agree that exception handling here is not needed for sub-process case.

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [reply torch/_inductor/select_algorithm.py:851]: @shunting314, sure I'll change. That's what I had initially, but as it turns out this is a nested function and the outer method _also_ has a param called 'choices'. So I didn't love the ambiguity. It would be easy to misspell 'options', for example, and be none-the-wiser because we'd silently reference the 'options' from the outer scope. I guess this is one of the reasons why I personally don't love nested functions.

File: torch/_inductor/select_algorithm.py
Comment (masnesral) [reply torch/_inductor/select_algorithm.py:860]: @shunting314 do you mean the overall parallel benchmark request in aggregate? Or do you mean each individual request? I wondered if the existing logging was already providing that. For example, I'm seeing this whenever I enable max autotune:

AUTOTUNE mm(512x768, 768x50265)
  triton_mm_1586 0.2621 ms 100.0%
  triton_mm_1585 0.2653 ms 98.8%
  triton_mm_1591 0.3040 ms 86.2%
  triton_mm_1587 0.3044 ms 86.1%
  triton_mm_1588 0.3108 ms 84.3%
  triton_mm_1584 0.3300 ms 79.4%
  triton_mm_1592 0.4615 ms 56.8%
  triton_mm_1594 0.4911 ms 53.4%
  triton_mm_1593 0.5753 ms 45.6%
  triton_mm_1590 0.6271 ms 41.8%
SubProcess AUTOTUNE takes 5.7510 seconds

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:366]: @eellison, yeah docs describe it better than I do: https://docs.python.org/3/library/pickle.html#pickle-picklable

> Note that functions (built-in and user-defined) are pickled by fully [qualified name](https://docs.python.org/3/glossary.html#term-qualified-name), not by value. [2](https://docs.python.org/3/library/pickle.html#id8) This means that only the function name is pickled, along with the name of the containing module and classes. Neither the function’s code, nor any of its function attributes are pickled. Thus the defining module must be importable in the unpickling environment, and the module must contain the named object, otherwise an exception will be raised.

So the crux of the test is replacing the benchmark() function with something simple than can exercise the obj.benchmark() call without actually doing the real benchmarking. The module unpickling needs to know that function definition.

**Comment by masnesral** [comment]
> Can you also share how much speedup we can get by leveraging multi-GPU.

@shunting314: here's a side-by-side comparison using the hf_Bart benchmark:

```
							      >	+ TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1
+ TORCHINDUCTOR_MAX_AUTOTUNE=1					+ TORCHINDUCTOR_MAX_AUTOTUNE=1
+ python benchmarks/dynamo/torchbench.py --device cuda --perf	+ python benchmarks/dynamo/torchbench.py --device cuda --perf
loading model: 0it [00:09, ?it/s]				loading model: 0it [00:09, ?it/s]
cuda eval  hf_Bart						cuda eval  hf_Bart
AUTOTUNE mm(512x768, 768x768)					AUTOTUNE mm(512x768, 768x768)
  mm 0.0148 ms 100.0%					      |	  mm 0.0141 ms 100.0%
  triton_mm_8 0.0154 ms 96.5%				      |	  triton_mm_8 0.0141 ms 99.5%
  triton_mm_3 0.0164 ms 90.8%				      |	  triton_mm_3 0.0161 ms 87.3%
  triton_mm_4 0.0171 ms 86.9%				      |	  triton_mm_4 0.0163 ms 86.4%
  triton_mm_9 0.0173 ms 85.6%				      |	  triton_mm_9 0.0167 ms 84.5%
  triton_mm_6 0.0175 ms 84.7%				      |	  triton_mm_6 0.0175 ms 80.3%
  triton_mm_5 0.0181 ms 82.2%				      |	  triton_mm_5 0.0179 ms 78.5%
  triton_mm_1 0.0197 ms 75.4%				      |	  triton_mm_1 0.0203 ms 69.5%
  triton_mm_2 0.0204 ms 72.7%				      |	  triton_mm_2 0.0203 ms 69.3%
  triton_mm_0 0.0300 ms 49.6%				      |	  triton_mm_0 0.0301 ms 46.8%
SingleProcess AUTOTUNE takes 5.1835 seconds		      |	SubProcess AUTOTUNE takes 4.7125 seconds
AUTOTUNE bmm(12x512x64, 12x64x512)				AUTOTUNE bmm(12x512x64, 12x64x512)
  triton_bmm_26 0.0138 ms 100.0%			      |	  triton_bmm_31 0.0132 ms 100.0%
  triton_bmm_25 0.0138 ms 99.5%				      |	  triton_bmm_26 0.0133 ms 99.8%
  triton_bmm_28 0.0146 ms 94.1%				      |	  triton_bmm_25 0.0136 ms 97.2%
  triton_bmm_31 0.0146 ms 94.1%				      |	  triton_bmm_32 0.0137 ms 96.5%
  triton_bmm_27 0.0147 ms 93.5%				      |	  triton_bmm_28 0.0139 ms 95.2%
  triton_bmm_32 0.0147 ms 93.5%				      |	  bmm 0.0148 ms 89.8%
  bmm 0.0148 ms 93.2%					      |	  triton_bmm_27 0.0154 ms 86.1%
  triton_bmm_24 0.0151 ms 90.9%				      |	  triton_bmm_24 0.0156 ms 84.7%
  triton_bmm_34 0.0164 ms 84.0%				      |	  triton_bmm_34 0.0169 ms 78.6%
  triton_bmm_33 0.0182 ms 75.4%				      |	  triton_bmm_33 0.0173 ms 76.5%
SingleProcess AUTOTUNE takes 4.6074 seconds		      |	SubProcess AUTOTUNE takes 1.0350 seconds
AUTOTUNE bmm(12x512x512, 12x512x64)				AUTOTUNE bmm(12x512x512, 12x512x64)
  triton_bmm_51 0.0154 ms 100.0%			      |	  triton_bmm_56 0.0141 ms 100.0%
  triton_bmm_56 0.0154 ms 99.6%				      |	  triton_bmm_51 0.0147 ms 96.3%
  triton_bmm_57 0.0155 ms 99.0%				      |	  triton_bmm_57 0.0154 ms 91.7%
  triton_bmm_54 0.0161 ms 95.4%				      |	  triton_bmm_53 0.0157 ms 90.0%
  triton_bmm_53 0.0162 ms 95.0%				      |	  triton_bmm_49 0.0160 ms 88.0%
  triton_bmm_52 0.0163 ms 94.5%				      |	  triton_bmm_52 0.0162 ms 87.0%
  bmm 0.0166 ms 92.3%					      |	  triton_bmm_54 0.0163 ms 86.6%
  triton_bmm_49 0.0174 ms 88.4%				      |	  bmm 0.0166 ms 85.0%
  triton_bmm_50 0.0175 ms 87.8%				      |	  triton_bmm_50 0.0179 ms 78.7%
  triton_bmm_48 0.0244 ms 62.8%				      |	  triton_bmm_48 0.0238 ms 59.3%
SingleProcess AUTOTUNE takes 4.2848 seconds		      |	SubProcess AUTOTUNE takes 0.7750 seconds
AUTOTUNE mm(512x768, 768x3072)					AUTOTUNE mm(512x768, 768x3072)
  mm 0.0214 ms 100.0%					      |	  mm 0.0206 ms 100.0%
  triton_mm_74 0.0230 ms 92.8%				      |	  triton_mm_74 0.0230 ms 89.9%
  triton_mm_73 0.0231 ms 92.6%				      |	  triton_mm_73 0.0238 ms 86.6%
  triton_mm_76 0.0251 ms 85.1%				      |	  triton_mm_76 0.0244 ms 84.6%
  triton_mm_75 0.0252 ms 85.0%				      |	  triton_mm_75 0.0246 ms 83.8%
  triton_mm_80 0.0296 ms 72.2%				      |	  triton_mm_80 0.0292 ms 70.6%
  triton_mm_72 0.0338 ms 63.3%				      |	  triton_mm_72 0.0328 ms 62.9%
  triton_mm_79 0.0377 ms 56.7%				      |	  triton_mm_79 0.0383 ms 53.9%
  triton_mm_82 0.0420 ms 50.9%				      |	  triton_mm_82 0.0416 ms 49.6%
  triton_mm_81 0.0465 ms 46.0%				      |	  triton_mm_81 0.0456 ms 45.3%
SingleProcess AUTOTUNE takes 4.7278 seconds		      |	SubProcess AUTOTUNE takes 0.8443 seconds
AUTOTUNE mm(512x3072, 3072x768)					AUTOTUNE mm(512x3072, 3072x768)
  mm 0.0277 ms 100.0%					      |	  mm 0.0283 ms 100.0%
  triton_mm_92 0.0358 ms 77.4%				      |	  triton_mm_92 0.0363 ms 78.0%
  triton_mm_87 0.0420 ms 66.1%				      |	  triton_mm_88 0.0414 ms 68.4%
  triton_mm_88 0.0420 ms 66.1%				      |	  triton_mm_87 0.0419 ms 67.6%
  triton_mm_89 0.0440 ms 63.0%				      |	  triton_mm_89 0.0440 ms 64.3%
  triton_mm_90 0.0451 ms 61.6%				      |	  triton_mm_90 0.0445 ms 63.6%
  triton_mm_93 0.0461 ms 60.2%				      |	  triton_mm_93 0.0462 ms 61.3%
  triton_mm_86 0.0563 ms 49.3%				      |	  triton_mm_86 0.0566 ms 50.1%
  triton_mm_85 0.0567 ms 49.0%				      |	  triton_mm_85 0.0566 ms 50.0%
  triton_mm_84 0.0708 ms 39.2%				      |	  triton_mm_84 0.0693 ms 40.9%
SingleProcess AUTOTUNE takes 4.5848 seconds		      |	SubProcess AUTOTUNE takes 1.1149 seconds
AUTOTUNE mm(512x768, 768x50265)					AUTOTUNE mm(512x768, 768x50265)
  triton_mm_1585 0.2616 ms 100.0%			      |	  triton_mm_1585 0.2570 ms 100.0%
  triton_mm_1586 0.2622 ms 99.8%			      |	  triton_mm_1586 0.2654 ms 96.8%
  triton_mm_1591 0.3036 ms 86.2%			      |	  triton_mm_1591 0.3005 ms 85.5%
  triton_mm_1587 0.3041 ms 86.0%			      |	  triton_mm_1587 0.3045 ms 84.4%
  triton_mm_1588 0.3084 ms 84.8%			      |	  triton_mm_1588 0.3071 ms 83.7%
  triton_mm_1584 0.3270 ms 80.0%			      |	  triton_mm_1584 0.3268 ms 78.6%
  triton_mm_1592 0.4618 ms 56.6%			      |	  triton_mm_1592 0.4638 ms 55.4%
  triton_mm_1594 0.4884 ms 53.6%			      |	  triton_mm_1594 0.4878 ms 52.7%
  triton_mm_1593 0.5923 ms 44.2%			      |	  triton_mm_1593 0.5705 ms 45.0%
  triton_mm_1590 0.6220 ms 42.1%			      |	  triton_mm_1590 0.6091 ms 42.2%
SingleProcess AUTOTUNE takes 5.7595 seconds		      |	SubProcess AUTOTUNE takes 1.0035 seconds
running benchmark: 100%|█████████████████████████████████████ |	running benchmark: 100%|█████████████████████████████████████
2.971x							      |	2.950x
```

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:246]: Went with lazy per conversation on that line of code

**Review by eellison — APPROVED** [review]
Looks good to me. let's get a review from @shunting314  as well.

Before we land this, we should make this off by default in the general case with a flag to make it enabled that is read from environment. 

For aot_compilation, I think we can turn this on. See: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L205-L209. cc @ipiszy, @aakhundov, was multi-device compile turned on by default for AIT ?

**Comment by aakhundov** [comment]
> was multi-device compile turned on by default for AIT ?

I guess, you mean multi-device profiling? Unless externally specified, AIT gets the devices to profile on from the `{CUDA,HIP}_VISIBLE_DEVICES}` environment variable. If that's not set, single device 0 is used for profiling (basically, serializing profiling of different kernel configurations). Here is the [relevant code](https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/compiler/compiler.py#L270-L275).

If you actually meant compilation, AIT uses all available CPUs for compilation by default. But that can also be overridden by the `NUM_BUILDERS` environment variable ([code](https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/backend/builder.py#L295-L297)).

**Review by shunting314 — APPROVED** [review]

**Comment by eellison** [comment]
> I guess, you mean multi-device profiling? Unless externally specified, AIT gets the devices to profile on from the {CUDA,HIP}_VISIBLE_DEVICES} environment variable. If that's not set, single device 0 is used for profiling (basically, serializing profiling of different kernel configurations)

Right, multi-device profiling. So yea, if it uses all of CUDA_VISIBLE_DEVICES, then multi-device profiling is on by default.

File: torch/_inductor/autotune_process.py
Comment (aakhundov) [inline torch/_inductor/autotune_process.py:193]: @masnesral @eellison I believe, there is a subtlety here worth revisiting.

Above we iterate over the `range(count)` and pass the `device` counter variable downstream, which is then set as the value of the `CUDA_VISIBLE_DEVICES` env variable when launching the subprocess. Now if the user has set `CUDA_VISIBLE_DEVICES=3,4,5` for the parent process, the `torch.cuda.device_count()` above will be 3, and we'll initialize the 3 subprocesses with `CUDA_VISIBLE_DEVICES=0`, `=1`, and `=2`. I guess, this is not what the user would expect?

In AITemplate, we actually parse the `CUDA_VISIBLE_DEVICES` env variable (fetched by `target.dev_select_flag()` [here](https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/compiler/compiler.py#L270-L275)) to detect the user-specified devices. Perhaps we should do something similar here, too? Or maybe there is a `torch.cuda.<something>` to detect the visible device IDs (didn't find from a quick check)?

File: torch/_inductor/autotune_process.py
Comment (aakhundov) [inline torch/_inductor/autotune_process.py:243]: Should there be a timeout? Can this get stuck forever otherwise?

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:193]: @aakhundov, ah sure. That sounds like a better behavior. Lemme work on a follow-up.

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:243]: @aakhundov, I'm not sure. Do you have a scenario in mind that would lead to a sub-process hanging? I'd love a repro to play around with. I'm new here and my naive assumption was that these kernels are fairly constrained and we wouldn't be concerned about arbitrary code. If the process crashes or exits unexpectedly, then we handle that scenario.

Comment (aakhundov) [reply torch/_inductor/autotune_process.py:243]: I don't have an exact scenario in mind. Just, given that we're relying on STDIN- / STDOUT-based communication, assuming that something can go wrong there leading to indefinitely long waiting. In AIT, we have a timeout for the whole profiling session (which, to be fair, we had to increase a couple of times due to longer e2e profiling that was initially expected).

**Pull request closed without merging**
